### DISCUSSION TITLE
How to subscribe within an action

### DISCUSSION THREAD

**Discussion by cortopy** [pmndrs/zustand#844] [Q&A]
I'm struggling to understand subscriptions within actions.  I have managed to make it work, but I'm facing deprecated methods and a confusing API.

I need to wait until a condition is met. In particular, I need the ability to know whether authentication has already happened

With a store like this:

```ts

type MyStore = {
  authenticated: boolean,
  authenticate: () => void,
  makeAuthenticatedRequest: () => void
}
```

I would like to write `makeAuthenticatedRequest` like this:

```ts
makeAuthenticatedRequest: async () => {
  console.log('fetching');

  await new Promise<void>(resolve => {
    const unsub = api.subscribe(state => {
      if (state.authenticated) {
        unsub();
        resolve();
      }
    });
  });

  console.log('authenticated! request can happen');  
  // and now carry on with logic
}
```

but the problem is that the subscription does not fire immediately. I have read the docs for `subscribeWithSelector`, which does have a `fireImmediately` option but since this is within an action, it's not a viable solution.

The following does work:

```ts
makeAuthenticatedRequest: async () => {
  console.log('fetching');

  await new Promise<void>(resolve => {
    const unsub = api.subscribe(
      () => {}, // adding something here (although it will never get called)
      state => {
        if (state.authenticated) {
          unsub();
          resolve();
        }
      });
  });

  console.log('authenticated! request can happen');  
  // and now carry on with logic
}
```

The problem with this is that the ts typings show it as deprecated. Also, the first argument is useless as it never gets called

However, this does fire immediately and with changes afterwards

This solution seems to work but also feels wrong

**Comment by dai-shi** [comment]
First, subscribing with an action might be possible, but it doesn't feel best approach because an action is not a way to get state.

If you see the deprecation message, it means the usage of `subscribeWithSelector` is somehow incorrect. In the first example, you are not using selector. In the second example, the selector does nothing, which means returning `undefined`.

Not 100% sure, but you want to do something like this?
```js
  await new Promise<void>(resolve => {
    const callback = state =>  {
      if (state.authenticated) {
        unsub();
        resolve();
      }
    };
    const unsub = api.subscribe(callback);
    callback(); // fire immediately
  });
```
I don't mean to recommend it.

**Comment by cortopy** [comment]
thanks @dai-shi! That worked, although with a small change

```
  await new Promise<void>(resolve => {
    const callback = state =>  {
      if (state.authenticated) {
        unsub();
        resolve();
      }
    };
    const unsub = api.subscribe(callback);
    callback(api.getState()); // CHANGE: pass state here
  });
```
When you say

> I don't mean to recommend it.

why not?

If it's because as you say `an action is not a way to get state`, every action has access to the `StoreGet` to get state synchronously anyway. Subscribing to state within an action may feel like an anti-pattern but it makes sense in an async action. After all, if one can always get state values synchronously, why not asynchronously? 

This is something I've always done with sagas and Redux observables, as it only makes sense within an async action. In fact, this pattern of waiting until a condition is met is now part of Redux toolkit too with their new listeners api for middleware [to replace what one used to do with sagas](https://github.com/reduxjs/redux-toolkit/releases/tag/v1.8.0). In the Redux ecosystem anything `async` has to happen outside reducers, but it was my understanding that in zustand async action handlers are ok. I've only started using zustand yesterday after many years of using Redux, so may be wrong.

Or you don't recommend it because of the way zustand works?

**Comment by dai-shi** [comment]
an action is not a way to get state to be used in react.

> it doesn't feel best approach

It's actually from different reason. sorry for the confusion.
Subscribing to entire state can be less efficient than awaiting for async function.
I think you have more controlled way to know when `authenticated` can be changed, other than watching the entire state, which can be very big.

That said, the library is rather unopinionated, and what works works. Unless you use values from actions in React (render functions), I don't think there are any pitfalls.

**Comment by cortopy** [comment]
That makes sense @dai-shi Your answer put me in the right track. I didn't understand `subscribeWithSelector` new middleware at first but then I had an eureka moment when I realised middleware in zustand can actually mutate the api

So I think I have a much better approach with a subscription:

```ts
export const authenticated$ = (
  callback: (selectedState: boolean, previousSelectedState: boolean) => void
) => {
  return useStore.subscribe(state => state.authenticated, callback, {
    fireImmediately: true
  });
};

export async function waitForAuthentication() {
  await new Promise<void>(resolve => {
    const unsub = authenticated$((authenticated: boolean) => {
      if (authenticated) {
        unsub();
        resolve();
      }
    });
  });
}
```

This only subscribes to the to value I need and can be used inside an action, or anywhere for that matter. I'm guessing this wouldn't have the performance penalty right?

I found an unexpected behaviour for this though, and opened [an issue](https://github.com/pmndrs/zustand/issues/848) for it

**Reply by dai-shi** [reply]
> middleware in zustand can actually mutate the api

Yeah, so technically, it can be extended so that we could do this:
```js
export async function waitForAuthentication() {
  await useStore.waitFor((state) => state.authenticated)
}
```
